proxy/http2/Http2DependencyTree.h

```diff
@@ -170,15 +214,63 @@ Tree<T>::find(uint32_t id)
   return n == nullptr ? nullptr : (n->is_shadow() ? nullptr : n);
 }
 
+template <typename T>
+void
+Tree<T>::add_ancestor(Node *node)
+{
+  if (node->parent != _root) {
+    _ancestors[ancestor_index].first  = node->id;
+    _ancestors[ancestor_index].second = node->parent->id;
+    ancestor_index++;
+    if (ancestor_index >= max_ancestors) {
+      ancestor_index = 0;
+    }
+  }
+}
+
+template <typename T>
+uint32_t
+Tree<T>::was_ancestor(uint32_t pid) const
+{
+  uint32_t i = (ancestor_index == 0) ? max_ancestors - 1 : ancestor_index - 1;
+  while (i != ancestor_index) {
+    if (_ancestors[i].first == pid) {
+      return _ancestors[i].second;
+    }
+    i = (i == 0) ? max_ancestors - 1 : i - 1;
+  }
+  return 0;
+}
+
 template <typename T>
 Node *
-Tree<T>::add(uint32_t parent_id, uint32_t id, uint32_t weight, bool exclusive, T t)
+Tree<T>::add(uint32_t parent_id, uint32_t id, uint32_t weight, bool exclusive, T t, bool shadow)
 {
-  Node *parent = find(parent_id);
+  Node *parent = find_shadow(parent_id); // Look for real and shadow nodes
+
   if (parent == nullptr) {
-    parent = add(0, parent_id, HTTP2_PRIORITY_DEFAULT_WEIGHT, false, nullptr);
+    if (parent_id > id) { // Parent is coming in the future, create a shadow
+      parent    = add(0, parent_id, HTTP2_PRIORITY_DEFAULT_WEIGHT, false, nullptr, true);
+      exclusive = false;
+    } else {
+      uint32_t pid = parent_id;
+      do {
+        pid = was_ancestor(pid);
+        if (pid != 0) {
+          parent = find(pid);
+        }
+      } while (pid != 0 && parent == nullptr);
+      if (parent == nullptr) {
+        // Found no ancestor, just add to root at default weight
+        weight    = HTTP2_PRIORITY_DEFAULT_WEIGHT;
+        exclusive = false;
+        parent    = _root;
+      }
+    }
   }
 
+  // If this has a pre-created shadow node, just fill it in and go
+  // Should we fix up the parent relationship?
```

Comment:
If we received STREAM frame with dependency field or PRIORITY frame, we should reprioritize the shadow node and make it regular node.